### DISCUSSION TITLE
Can VDH skip non-video lectures when it's creating/downloading filenames?

### DISCUSSION THREAD

**Discussion by rihntintin** [aclap-dev/video-downloadhelper#812] [Help and Q&A]
### Your question

Take this scenario: When there are some text written lectures in a course, instead of video lectures, is there a way to inform VDH to skip those text lectures when it's creating the next video filename? E.G. when lecture 4 is a text lecture, and VDH has downloaded these 3 files: Python.mp4; Python-01.mp4; Python-02.mp4. The next file is a text lecture, so I want to skip it and download the next video lecture to Python-04.mp4, but VDH will name the video as Python-03.mp4.

This past March and April 2024, I noticed when I rename the file from Python-03.mp4 to Python-04.mp4, VDH would correctly name the next file as Python-05.mp4 and all the subsequent files were correctly named. Even when there was other text lectures, I was able to rename the file to the bypass each text lecture, and VDH understood this and correctly name all subsequent filenames.

But then in May of 2024, VDH has changed back to pre-late March functionality, where I would have to create a zero byte file with the name Python-03.mp4 in order to have VDH create each subsequent video filenames correctly.

Thank you in advance! I love the new features of VDH. It's THEE greatest app/extension for me. I use it all the time. I can't live without VDH!

- Pete

### Addon & CoApp version and details

version: 9.0.2.3
target: google
channel: stable
lang: en-US
coapp: {"found":true,"new_version":false,"path":"/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp","version":"2.0.19"}
license: {"accepted":true,"email":"<removed>","key":"1ced0282b1ab4aff"}
platform: x86-64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36
{
    "valid_license_message_has_been_displayed": true,
    "used_sidebar_button": true,
    "used_popup_button": true,
    "download_directory": "/Users/bonnieoliverimac2017/Movies/Udemy",
    "view_options": {
        "all_tabs": false,
        "hide_downloaded": false,
        "low_quality": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "sort_by_status": true,
        "sort_reverse": false
    },
    "use_sidebar": false,
    "open_count_store": 34,
    "successfull_dl": 14
}

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I'm just another user.

I'm a bit confused by your problem description.  Here is how I understand what you are trying to do.  Do correct me if I'm wrong.  I believe you are seeing a scenario in which you download 3 videos & VDH downloads them successfully & automatically gives them names you like.  The fourth item on the page is not even a video so you're not using VDH on that item.  The fifth item on the page is the 4th video.  But you are doing a Download As to change its name so it's Video-05.  This completes successfully.  Are you saying your 5th download, which is object 6 on the page, & you are doing this one after completing Video-05 successfully, is not automatically getting the name with -06 in it?  If the web page in question does not require a registration, perhaps you could give us the URL.  Don't bother if it requires a registration.  In either case, a few screenshots illustrating your process might clarify things.

**Comment by rihntintin** [comment]
Hi Grampa, Thank you for your question & assistance. First, I forgot to mention that VDH is creating these filenames itself. The actual filenames that VDH uses are the entire course name, plus a '-' and then the 2 digit number, in sequence. The Download As feature doesn't work. It will always resort back to the default name (i.e. the course name) when I try to download the next file. Meaning, it will not reuse the name I typed in. And more importantly, when I use the Download As feature, it will not automatically include the 2 digits in sequence, so there's no point in using Download As and it's not helpful for me. (but that's for a different Q&A session).

Anyway, let me try a different way of describing the issue.

After VDH has correctly creating Python.mp4, Python-01.mp4 & Python-02 and I skip the text lecture, VDH will create Python-03.mp4, but then I have to manually rename this file to Python-04.mp4.  But then for the next video lecture, VDH will recreate the file as Python-03.mp4 and then I have to manually rename the file to Python-05.mp4. The next video lecture VDH will recreate the file as Python-03.mp4 and then I have to manually rename the file to Python-06.mp4.

But during late March and April, when VDH would create the Python-03.mp4 the first time, I would rename it to Python-04.mp4 and VDH would "see" this new -04 filename and then "VDH knew" to skip the -03 file and all other files would have their correct filenames and the correct 2 digits in sequence. This was perfect and a wonderful new feature, but now it's gone way in the May update to VDH.

I hope this helps. I don't know how screenshots would help, but I'm open to any suggestions. And I'll try to think of a better way to describe this issue. Thank you again for trying to me. I truly appreciate it and I will respond to any new suggestions in order to clarify this issue.

For the past 10+ years, my work around is to create an empty file with the Python-03.mp4 name and VDH works fine until there is another text lecture. In which case, I just create another empty file to force VDH to keep the sequence correct.

- Pete

**Reply by GrampaWildWilly** [reply]
I think I've got it now.  VDH keeps seeing the discontinuity at -03 (or those subsequent gaps) & keeps trying to fill it despite there being higher numbers already in the directory.  It's interesting that Download As doesn't suit your purposes.  Are you saying that with each download, it keeps proposing the same file name repeatedly?  That sounds like it might be a regression that @paulrouget Paul should look into.  I have never dealt with a page like the one you're describing.  I gather that your omission of a URL means it's a page that needs a registration.  It sounds like it would be a great test case but for that.

**Reply by GrampaWildWilly** [reply]
I would have proposed doing something like this:

Download Python
Download Python-01
Download Python-02
Download As Python-04
Download Python-05

But you say Download As doesn't recognize the file name collision.  It doesn't propose Python-03 so that you can manually change it to -04.  That does not sound right.  I know VDH will nag you if you go to the next step & launch the download.  It recognizes then that you have a name collision & prompts you with a Yes/No nag whether to replace the target file.  But that's a step too late.  Even Download As should recognize in the first step that the name it's proposing is a collision & it should append the -xx.  That's how it used to work, as I recall.  Download As should propose the -xx names, same as Download does automatically.  Providing a dialog that allows the user to change the name does not remove the responsibility from VDH to recognize a name collision up front.  Its first proposed name should include the -xx suffix on a name collision.  Thinking back on some Download As functions I've done, when there is a name collision, I haven't seen the -xx suffix.  Since I don't do the kind of downloads in a series like you're doing, I haven't thought twice about it.  I just changed the file name to avoid the collision.  Or I said Yes & let it replace the file I already have.  So now I definitely agree with you.  Paul needs to fix that.  That is a regression.

For the skip, it seems to me that there must be code in the product that looks for gaps on purpose.  That seems very intentional, that it was designed quite carefully to do that.  I would propose a user setting to affect that behavior.  But I think it would be too hard to give it a meaningful name that people would understand.  I think there's no harm in discontinuities in the numbering.  Those would happen only in a case like yours where you actually want the discontinuities.  I say just detect the highest number in a -xx suffix, add 1, & propose that.  Don't go to the trouble of looking for the gaps.  Let the user deal with gaps manually after the fact if he doesn't like them.  In most cases, there wouldn't be any gaps.  You've proposed an interesting test case.  I say remove the code that looks for discontinuities.

**Reply by rihntintin** [reply]
Yes, I forgot to mention, it's an Udemy.com course. Linkedin.com also works the same way.
If you can suggest another site to test, I will try it.
Yes, Download As only uses the course name as the default.
But when I enter a new name, VDH will create the file using that new name, but when I want to download a new file, it doesn't remember what I had just typed in. Therefore, I have to re-enter the new name again.

Here is something new I just discovered: I just noticed an option called "Remember as default action", but I don't really want to change the filename, per se, I just want to easily skip the 2 digit sequence for non-video lectures.

Thanks again for your reply.

To summarize, can one of the developers reinstate the previous feature from the April release, where VDH reads my current folder location and finds the filename with the largest 2 digit sequence and increases the 2 digits by 1 before creating the next filename?

In order words, VDH sees Python-04.mp4 and then creates the next file as Python-05.mp4 and therefore VDH doesn't create any new file with the Python-03.mp4 name. Thereby, skipping the text lecture, which would have been the Python-03.mp4 file.

**Reply by GrampaWildWilly** [reply]
Udemy?  You have to write a tutorial for all the people reporting their inability to use VDH on that site.  There's perhaps hundreds of reports of problems there & on Hotmart if you combine this forum with our old one on Google.

To summarize even more, don't look for gaps in the sequence numbers.  And do append -xx sequence numbers on Download As functions, same as VDH always used to in past releases.

The "Remember as default action" controls the default you see on the action button on the main VDH menu.  It doesn't really address your issue of file naming at all.  They're separate areas of discussion.

**Reply by rihntintin** [reply]
I somewhat agree about Udemy, but now that I have a newer iMac, all my old Udemy issues when away. And now VDH is great with Udemy! If others are using older computers and they see issues with Udemy, maybe they just need a new computers (if they can afford it). My new iMac resolved my issues.

I would suggest you and all the other support people and non-official people who are helping the users, check if the users are using older computers, older Operating Systems, and/or older browsers, then gently suggest to them to upgrade their computers. Good Luck with that. lol

BTW: Thank you to ALL the VDH developers and ALL who help users. You guys and gals are great!!!! I truly appreciate all of you!!!!

As for your summarize paragraph, I'm guessing if I activate the "Remember as default action" then it will be saved in the default Download button and therefore, the -xx sequence would be incorporated in all filenames after that. But I don't want to fool around with it, since my work around is good enough for me.

**Reply by GrampaWildWilly** [reply]
No.  That's not what "Remember as default action" does.  In #665 while discussing something else, I happened to post this image:

![336282225-5a3428fa-f092-4ae1-bc2e-545afebb7c1e](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/6487fa33-edcf-435c-85b8-2aeac91a6e79)

It explains how to use "Remember as default action" & what it does.  It does not affect file naming at all.  They are 2 completely unrelated topics.  "Remember as default action" is just a convenience that allows you to skip the secondary VDH menu & execute your desired action from the main VDH menu.  It has no effect on the naming of files.  You might read that other discussion to see the image in context.  It might clarify things.  But you should stop trying to connect the default action with your file naming thing.  You are seeing renaming a file as an "action."  The actions in VDH parlance are Download, Download As, Download Audio, etc., the actions you can see on the secondary VDH menu in that image.  There's nothing there about file renaming.  You can experiment with "Remember as default action" to your heart's content.  You will see that your file renaming issue will be unaffected.

OK.  So you're using Mac.  I wonder if being current on Windows would have the same impact.  Well, you've posted your advice here.  We can only refer Udemy & Hotmart users to your advice when (not if, there's going to be more folks reporting this) they come in here with their problems.

**Reply by GrampaWildWilly** [reply]
Let me quote this again, because it isn't really directed at you but at @paulrouget Paul:

> To summarize even more, don't look for gaps in the sequence numbers. And do append -xx sequence numbers on Download As functions, same as VDH always used to in past releases.

This is me suggesting to Paul what he needs to do to the code of VDH.  This is not advice for users about how to use VDH.

**Reply by rihntintin** [reply]
Thank you, Grampa, for your detailed clarification. I truly appreciate all you're doing for my question and for all the other user's questions you work on.

**Reply by rihntintin** [reply]
Hey @GrampaWildWilly, that issue of skipping a -xx, is now working.
The following may not be related to the -xx issue, but I wanted to include it. I always download in 720p. The last time I downloaded in 720p was May 22 at 3:48pm. The next time I download a lecture was on May 23 at 12:15am and it was in 1080p. I didn't make any changes. VDH must have changed it on its own, as far as I'm concerned. Then, last night, I noticed that my lectures were being saved in 1080p. That may or may not be related in the -xx issue.

Last night, I tried to switch the downloads back to 720p, by clicking on 720p and then clicking on the "Always prefer this quality". But it only worked for that single lecture. For the next lecture, VDH reverted the default setting back to 1080p. Then I click on 720p and "Always prefer this format". Same thing. It only worked for that 1 download and reverted back to 1080p as the default. I click o 720p and click both quality & format, & same thing happened.

I figured since 720p was above the 2 other options, that it was the correct process. Today, I decided to click both quality & format options and then click on 720p and it worked for all downloads & the 720p is still the default.

Then I download 2 lectures from a new course. I renamed the 2nd filename from -01 to -04. The next lecture was saved as -05. The next lecture was saved as -06 and the next as -07; next as -08 and next as -09. Great! it's working.

Then I renamed the -09 to -19 but the next lecture was saved as -10; next -11; next as -12. So, I renamed the -19 back to -09 and the next lecture was saved as -13. I renamed -13 to -14 and the next lecture was saved as -15. So, it's working correctly, but you can't skip 10 lectures, but you can skip 1 or 2 lectures. I hope I explained it without confusing anyone.

**Reply by GrampaWildWilly** [reply]
I'm not sure but it sounds to me like it's not working consistently.  Only Paul can untangle this.  I'm just another user, like you.  I suppose if you like the way it's working now, we shouldn't tempt fate & dig into it any deeper.

**Reply by rihntintin** [reply]
The inconsistency with -xx is not a big deal for Paul to ever fix it. It can be the last item on his todo list.

Just now I was downloading an Udemy course, I stopped before downloading the -22 lecture, for a few minutes, to speak to my wife, and then click on the download. I received a message (something, like) apologizing for not being able to download that format. I deleted the message and click on download and received the same message. I then refresh the Chrome tab and clicked download and it successfully download the video, but with a -24 filename, as if VDH had considered 2 previous error attempts as valid downloads and therefore skipped -22 and -23 files.

I mention, since it may help @paulrouget to better understand how the logic is or isn't working. I hope it helps him.

I can never say it enough, how TRULY WONDERFUL VDH is for me and how great it is to create mp4 files to play offline, without being stuck using Udemy's very limited mp4 player. Especially, since VLC is vastly superior to QuickTime Player Udemy's website.

Something that is rarely mentioned in online chats, the volume of mp4 files can be increased to 200% the volume level (using VLC) from the maximum volume level on QuickTIme or any other mp4 player app. Which is great since some Udemy instructors have the volume down so low you can barely hear them on MacOS' normal 100% level.

**Reply by GrampaWildWilly** [reply]
Sounds like you hit a timeout.  Pages all over the web time out.  YouTube does.  I've visited other sites that do.  Some time out in minutes.  Some in hours.  But lots of sites time out.  The Metropolitan Opera has a rather generous timeout window.  Their content being operas, I think their timeout delay is something like 6 or 8 hours.  But they do time out.  I guess Udemy is on the list.  You should probably reload the page after just 1 failure like the one you got.  Were there residual bogus files with -22 & -23 in their names in your target directory?

Volume is a tricky thing.  I've got my computer hooked up to my surround sound system so my volume control is outside my computer.  I just set volume to 100% on everything in the computer & do the real volume control on the receiver.  If it's possible for you (or anybody) to set it up like that, it's really the way to go.

**Reply by rihntintin** [reply]
I forgot to mention, no there weren't any -22 or -23 files created. That may sound really bizarre, but I know from software projects, that programming & logic can get mixed up from what the developer had intended. But once again, this issue is small stuff, IMHO. I feel this discussion should be closed and not waste Paul's time with.

**Reply by GrampaWildWilly** [reply]
Oh no.  This is a bug & it needs to be fixed.  At what priority is a different question.  But this definitely needs to be fixed.  As you discovered in the other thread you found, Paul went to the trouble of adding this logic.  It wasn't in the earlier VDH 9 releases.  So it's new code.  It needs to be fixed . . . eventually.

The absence of those 2 files is actually what I expected.  The downloads failed so any files it may have created would get cleaned up.  That would probably have been another bug if the 2 files had existed.
